### PR TITLE
feat(versioning/ubuntu): Support for Ubuntu codenames

### LINKED ISSUE #13645: Support for Ubuntu codenames
### What would you like Renovate to be able to do?

Renovate should understand Ubuntu release codenames.

It should e.g. know that bionicequals v18.04 and offer upgrading it to focal.

### If you have any ideas on how this should be implemented, please tell us here.

Implement codenames as ranges like we've done in nodejs

ref:
- #13466
- #13450

/cc @fgreinacher

### Is this a feature you are interested in implementing yourself?

Maybe

### PR DESCRIPTION
**Pull request by Gabriel-Ladzaretti**
## Changes
Added a middleware to handle codenames.
Current logic is left untouched.

## Context
Closes #13645

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by Gabriel-Ladzaretti** [comment]
note: will probably move the hardcoded data to a file under renovate/data.

File: lib/modules/versioning/ubuntu/index.ts [hunk 1]
```diff
@@ -6,15 +6,81 @@ export const displayName = 'Ubuntu';
 export const urls = ['https://changelogs.ubuntu.com/meta-release'];
 export const supportsRanges = false;
 
+const distroInfo = `Ubuntu 4.10 "Warty Warthog"
```
Comment (viceice) [inline lib/modules/versioning/ubuntu/index.ts:9]: Please convert to static json (you can create a js file to tools for it) and safe to data folder.
Then our `generate-imports` tool should automatically add it to the generated files.
checkout nodejs schedule file usage for sample

https://github.com/renovatebot/renovate/blob/4f65b57225a42f9deb010f8dbfeb06eb94c0dd69/tools/generate-imports.mjs#L66

https://github.com/renovatebot/renovate/blob/dca3418bbdbf1024ac6807e591562b4042bc846d/lib/modules/versioning/node/schedule.ts#L1

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/versioning/ubuntu/index.ts (hunk 1 above)
Comment (Gabriel-Ladzaretti) [reply lib/modules/versioning/ubuntu/index.ts:9]: thanks, exactly what i was aiming at 😄

File: lib/modules/versioning/ubuntu/index.ts [hunk 2]
```diff
@@ -6,15 +6,81 @@ export const displayName = 'Ubuntu';
...
+Ubuntu 5.04 "Hoary Hedgehog"
```
Comment (JamieMagee) [inline lib/modules/versioning/ubuntu/index.ts:10]: Is `Ubuntu` not a bit redundant?

File: lib/modules/versioning/ubuntu/index.ts [hunk 3]
```diff
@@ -6,15 +6,81 @@ export const displayName = 'Ubuntu';
...
+Ubuntu 22.04 LTS "Jammy Jellyfish"`;
+
+const codenameToVersion = new Map<string, string>();
+const versionToCodename = new Map<string, string>();
+
+// const populateMaps = (match): string =>
+//   match.replace(/(\r?\n)/g, '$1// renovate-replace ');
+//
+// distroInfo.replace(/([\d]*\.\d*)* "(.*)/g, populateMaps);
+const regex = /([\d]*\.\d*)( LTS)? "(.*) /g;
```
Comment (JamieMagee) [inline lib/modules/versioning/ubuntu/index.ts:53]: You can constrain this a little more:
```suggestion
const regex = /(\d{1,2}\.(?:04|10))(?: LTS)? "([\w\s])+"/g;
```

(This also changes the capture groups so ` LTS` isn't captured)

File: lib/modules/versioning/ubuntu/index.ts (hunk 2 above)
Comment (Gabriel-Ladzaretti) [reply lib/modules/versioning/ubuntu/index.ts:10]: well yes, but then again its auto generated.
other than that ill be using the [following](https://github.com/renovatebot/renovate/issues/13645#issuecomment-1065063748).

File: lib/modules/versioning/ubuntu/index.ts [hunk 4]
```diff
@@ -111,7 +125,14 @@ function minSatisfyingVersion(
 }
 
 function getNewValue(newValueConfig: NewValueConfig): string {
-  return newValueConfig.newVersion;
+  let newVer = newValueConfig.newVersion;
+  if (isCodename(newValueConfig.currentValue)) {
+    if (isCodename(newVer)) {
```
Comment (PhilipAbed) [inline lib/modules/versioning/ubuntu/index.ts:130]: combine the Ifs with &&?

Comment (PhilipAbed) [inline lib/modules/versioning/ubuntu/index.ts:130]: isCodename(newValueConfig.currentValue) && !isCodename(newVer)

File: lib/modules/versioning/ubuntu/distribution.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  const codename = ubuntuDistroInfo[version];
+  versionToCodename.set(version, codename);
+  codenameToVersion.set(codename, version);
+}
+
+export function isCodename(input: string): boolean {
+  return !!codenameToVersion.get(input);
+}
+
+export function getVersionByCodename(input: string): string {
```
Comment (PhilipAbed) [inline lib/modules/versioning/ubuntu/distribution.ts:27]: why do we have conversion maps?
if we detect a codename, we update a codename, 
if we have a version we update a version?

Comment (Gabriel-Ladzaretti) [reply lib/modules/versioning/ubuntu/distribution.ts:27]: because the underlining logic works with numeric versions only, thats what we are getting from the datasource.

File: lib/modules/versioning/ubuntu/distribution.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+const versionToCodename = new Map<string, string>();
+
+for (const version of Object.keys(ubuntuDistroInfo)) {
+  const codename = ubuntuDistroInfo[version];
+  versionToCodename.set(version, codename);
+  codenameToVersion.set(codename, version);
+}
+
+export function isCodename(input: string): boolean {
+  return !!codenameToVersion.get(input);
```
Comment (PhilipAbed) [inline lib/modules/versioning/ubuntu/distribution.ts:24]: codenameToVersion.contains(input) ?

File: lib/modules/versioning/ubuntu/distribution.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
+import dataFiles from '../../../data-files.generated';
+
+export type UbuntuDistroInfo = Record<string, string>;
+
+// Data file generated with:
+// ubuntu-distro-info --all -f | sed -r 's/Ubuntu|"|LTS //g; s/([0-9]+.[0-9]+) /\1=/; s/.*/\L&/; s/(=[a-z]*) [a-z]*/\1/g; s/^[ \t]*//' | jo
+const dataFile = 'data/ubuntu-distro-info.json';
+
+const ubuntuDistroInfo: UbuntuDistroInfo = JSON.parse(
```
Comment (MaronHatoum) [inline lib/modules/versioning/ubuntu/distribution.ts:9]: what if json file does not exist? wrong path?

File: lib/modules/versioning/ubuntu/distribution.ts (hunk 6 above)
Comment (Gabriel-Ladzaretti) [reply lib/modules/versioning/ubuntu/distribution.ts:24]: done

File: lib/modules/versioning/ubuntu/distribution.ts (hunk 7 above)
Comment (Gabriel-Ladzaretti) [reply lib/modules/versioning/ubuntu/distribution.ts:9]: you are right, "dataFIle" is misleading, as its is not referring to an actual fs file. but rather a key used to index into an auto generated map.

CI builds will fail if the file is missing, so its part of the project. so i guess we assume it exists.

**Comment by Gabriel-Ladzaretti** [comment]
sample renovate bot pr:
https://github.com/ladzaretti/docker-testing-1/pull/1

File: lib/modules/versioning/ubuntu/distribution.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
+import dataFiles from '../../../data-files.generated';
+
+export type UbuntuDistroInfo = Record<string, string>;
+
+// Data file generated with:
+// ubuntu-distro-info --all -f | sed -r 's/Ubuntu|"|LTS //g; s/([0-9]+.[0-9]+) /\1=/; s/.*/\L&/; s/(=[a-z]*) [a-z]*/\1/g; s/^[ \t]*//' | jo
+const ubuntuJsonKey = 'data/ubuntu-distro-info.json';
+
+const ubuntuDistroInfo: UbuntuDistroInfo = JSON.parse(
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (viceice) [inline lib/modules/versioning/ubuntu/distribution.ts:10]: Why we need this? 🤔 Shouldn't be required.

File: lib/modules/versioning/ubuntu/distribution.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  dataFiles.get(ubuntuJsonKey)!
+);
+
+const codenameToVersion = new Map<string, string>();
+const versionToCodename = new Map<string, string>();
```
Comment (viceice) [inline lib/modules/versioning/ubuntu/distribution.ts:15]: Isn't `ubunuDistroInfo` the same? We should hold unused object on module scope.

**Review by viceice — COMMENTED** [review]
I think we should handle cod names as range, like we do on node versioning.

File: lib/modules/versioning/ubuntu/distribution.ts (hunk 8 above)
Comment (Gabriel-Ladzaretti) [reply lib/modules/versioning/ubuntu/distribution.ts]: removed it

File: lib/modules/versioning/ubuntu/distribution.ts (hunk 9 above)
Comment (Gabriel-Ladzaretti) [reply lib/modules/versioning/ubuntu/distribution.ts]: removed that as well :)

File: data/ubuntu-distro-info.json [hunk 10]
```diff
@@ -0,0 +1,38 @@
+{
```
Comment (HonkingGoose) [inline data/ubuntu-distro-info.json:1]: This cool new feature needs to be documented. 😉 

The documentation should explain:
- That we now support Ubuntu codenames.
- What's the correct format for the codenames: 
  - Do I need to use the full codename like `Focal Fossa` or can I use a part of the codename? Like just `focal` or `fossa`?

Comment (Gabriel-Ladzaretti) [reply data/ubuntu-distro-info.json:1]: The data source for tags/versions here is the docker one.
did a little research and basically the versioning is case sensitive (i.e. `FROM ubuntu:<codename>`) and requites only the first term of the codename phrase.

**Comment by Gabriel-Ladzaretti** [comment]
A side effect of this feature is that when filtering the response from the docker datasource, we are also getting back the 
codenames (`focal`, `bionic` and so on) as valid tags as well, are we ok with that?

in contrast, currently we are only getting the numeric tags as response.

On another note, 
do we have some other manager other than docker who uses this? couldn't fine another one

**Comment by Gabriel-Ladzaretti** [comment]
> I think we should handle cod names as range, like we do on node versioning.

ubuntu versioning does not support ranges nor docker , as opposed to npm who does and is used in the node one.
how would you propose to handle ranges here :?

**Comment by rarkins** [comment]
Re: ranges, I can't think of a use case for it right now, so no need to add it purely for this requirement.

**Comment by viceice** [comment]
> Re: ranges, I can't think of a use case for it right now, so no need to add it purely for this requirement.

I think if we don't do this. it can cause update flip-flop from version to codename and back to version.

**Comment by Gabriel-Ladzaretti** [comment]
> I think if we don't do this. it can cause update flip-flop from version to codename and back to version.

This change might reduce some of the flip-flops - 
https://github.com/renovatebot/renovate/blob/762749a5383f0b1f34509f96011af2069d75aae1/lib/modules/versioning/ubuntu/index.ts#L126-L137

File: docs/usage/docker.md [hunk 11]
```diff
@@ -117,6 +117,26 @@ If you wish to enable major versions then add the preset `docker:enableMajor` to
...
 For example:
 
+### Ubuntu codenames
+
+Renovate understands [Ubuntu release codename](https://wiki.ubuntu.com/Releases) tags and will offer upgrading them to the latest LTS release available (e.g. from `ubuntu:xenial` to `ubuntu:focal`) as long as it is inline with the following naming scheme:
+
+- First term of the full codename is used (e.g. `bionic` for `Bionic Beaver` release)
+- Codename is all lowercase
+
+####Example:<br>
```
Comment (HonkingGoose) [inline docs/usage/docker.md]:
```suggestion
For example, Renovate will offer to upgrade the following `Dockerfile` layer:
```

We can drop the `h4` and rewrite the sentence instead. 😉

File: docs/usage/docker.md [hunk 12]
```diff
@@ -117,6 +117,26 @@ If you wish to enable major versions then add the preset `docker:enableMajor` to
...
+Renovate will offer to upgrade the following `Dockerfile` layer
```
Comment (HonkingGoose) [inline docs/usage/docker.md:129]:
```suggestion
```

File: docs/usage/docker.md [hunk 13]
```diff
@@ -117,6 +117,26 @@ If you wish to enable major versions then add the preset `docker:enableMajor` to
 Renovate has some Docker-specific intelligence when it comes to versions.
 For example:
 
+### Ubuntu codenames
+
+Renovate understands [Ubuntu release codename](https://wiki.ubuntu.com/Releases) tags and will offer upgrading them to the latest LTS release available (e.g. from `ubuntu:xenial` to `ubuntu:focal`) as long as it is inline with the following naming scheme:
```
Comment (HonkingGoose) [inline docs/usage/docker.md:122]:
```suggestion
Renovate understands [Ubuntu release codename](https://wiki.ubuntu.com/Releases) tags and will offer upgrading them to the latest LTS release available (e.g. from `ubuntu:xenial` to `ubuntu:focal`).
For this to work you must follow this naming scheme:
```

Split up the sentence into smaller parts, and rewrite the platform for the list.

File: docs/usage/docker.md [hunk 14]
```diff
@@ -117,6 +117,26 @@ If you wish to enable major versions then add the preset `docker:enableMajor` to
...
+
+- First term of the full codename is used (e.g. `bionic` for `Bionic Beaver` release)
```
Comment (HonkingGoose) [inline docs/usage/docker.md:124]:
```suggestion
- The first term of the full codename is used (e.g. `bionic` for `Bionic Beaver` release)
```

File: docs/usage/docker.md [hunk 15]
```diff
@@ -117,6 +117,26 @@ If you wish to enable major versions then add the preset `docker:enableMajor` to
...
+- Codename is all lowercase
```
Comment (HonkingGoose) [inline docs/usage/docker.md:125]:
```suggestion
- The codename is in lowercase
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Comment by Gabriel-Ladzaretti** [comment]
had a issue with the linter, sorry about that 😅

File: docs/usage/docker.md [hunk 16]
```diff
@@ -117,6 +117,27 @@ If you wish to enable major versions then add the preset `docker:enableMajor` to
 Renovate has some Docker-specific intelligence when it comes to versions.
 For example:
 
+### Ubuntu codenames
+
+Renovate understands [Ubuntu release codename](https://wiki.ubuntu.com/Releases) tags and will offer upgrading them to the latest LTS release available (e.g. from `ubuntu:xenial` to `ubuntu:focal`).
```
Comment (HonkingGoose) [inline docs/usage/docker.md:122]:
```suggestion
Renovate understands [Ubuntu release code names](https://wiki.ubuntu.com/Releases) and will offer upgrades to the latest LTS release (e.g. from `ubuntu:xenial` to `ubuntu:focal`).
```

- We can drop "tags" and just use "code names".
- Replace `upgrading them` with `upgrades`.
- People won't expect Renovate to update things that "are not available". So we can drop the "available" part. 😉

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/modules/versioning/ubuntu/distribution.ts [hunk 17]
```diff
@@ -0,0 +1,39 @@
+import dataFiles from '../../../data-files.generated';
+
+export type UbuntuDistroInfo = Record<string, string>;
+
+// Data file generated with:
+// ubuntu-distro-info --all -f | sed -r 's/Ubuntu|"|LTS //g; s/([0-9]+.[0-9]+) /\1=/; s/.*/\L&/; s/(=[a-z]*) [a-z]*/\1/g; s/^[ \t]*//' | jo
```
[the remark this answers is not in the document]
Comment (HonkingGoose) [reply lib/modules/versioning/ubuntu/distribution.ts:6]: @Gabriel-Ladzaretti Can you move the commented shell code into a `.sh` file?

[the remark this answers is not in the document]
Comment (Gabriel-Ladzaretti) [reply lib/modules/versioning/ubuntu/distribution.ts:6]: @HonkingGoose 
is this suggestion related to the build tools or to this pr?
i mean i couldnt find any bash scripts here in renovate.

[the remark this answers is not in the document]
Comment (HonkingGoose) [reply lib/modules/versioning/ubuntu/distribution.ts:6]: My suggestion is about this PR. 😉
We have some utility scripts, but they're TypeScript files: https://github.com/renovatebot/renovate/tree/main/tools

I think a TypeScript file might work. But I don't know if you can put a Bash script into a TypeScript file. So you'll have to experiment a bit.

@viceice Do you have any tips?

[the remark this answers is not in the document]
Comment (Gabriel-Ladzaretti) [reply lib/modules/versioning/ubuntu/distribution.ts:6]: sure, where should i place it? couldnt find any other shell script here in renovate

[the remark this answers is not in the document]
Comment (Gabriel-Ladzaretti) [reply lib/modules/versioning/ubuntu/distribution.ts]: > @Gabriel-Ladzaretti Can you move the commented shell code into a `.sh` file?
Done.

Im just not quite sure as to where to place it. for now its in `lib/modules/versioning/ubuntu/`

[the remark this answers is not in the document]
Comment (viceice) [reply lib/modules/versioning/ubuntu/distribution.ts]: Would like to have it as jsm file, so it works on any platform, like or other scripts in `tools/` folder. There's also the right place for it.

File: lib/modules/versioning/ubuntu/ubuntu-distro-json-generate.sh [hunk 18]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
```
Comment (HonkingGoose) [inline lib/modules/versioning/ubuntu/ubuntu-distro-json-generate.sh:1]: Copied and adjusted the comment from @viceice so it can be tracked and resolved later. 😉 

> Would like to have it as `.mjs` file, so it works on any platform, like our other scripts in the `tools/` folder. That's also the right place for it.

Comment (Gabriel-Ladzaretti) [reply lib/modules/versioning/ubuntu/ubuntu-distro-json-generate.sh]: mjs file added 😄

**Review by HonkingGoose — DISMISSED** [review]
I'm happy with the docs.

**Review by rarkins — DISMISSED** [review]

File: tools/ubuntu-distro-json-generate.mjs [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+
+if (!shell.which(`sed`)) {
+  shell.echo('This script requires sed, exiting...');
+  shell.exit(2);
+}
+
+if (!shell.which(`jo`)) {
+  shell.echo('This script requires jo, exiting...');
+  shell.exit(2);
+}
```
Comment (JamieMagee) [inline tools/ubuntu-distro-json-generate.mjs:19]: nit: I would see if we can do more of this in Node as I don't think `jo` is a common default package. But this can be handled at a later date.

**Review by JamieMagee — DISMISSED** [review]
Nice! Now I guess we need to do this for Debian too 😅

File: lib/modules/versioning/ubuntu/distribution.ts [hunk 20]
```diff
@@ -0,0 +1,39 @@
+import dataFiles from '../../../data-files.generated';
+
+export type UbuntuDistroInfo = Record<string, string>;
+
+// Data file generated with:
+// ubuntu-distro-json-generate.mjs
+const ubuntuJsonKey = 'data/ubuntu-distro-info.json';
+
+const ubuntuDistroInfo: UbuntuDistroInfo = JSON.parse(
+  dataFiles?.get(ubuntuJsonKey) ?? ''
```
Comment (viceice) [inline lib/modules/versioning/ubuntu/distribution.ts:10]: 🤔 do we really need those fallbacks? Error should be caught on ci before publish

File: tools/ubuntu-distro-json-generate.mjs (hunk 19 above)
Comment (viceice) [reply tools/ubuntu-distro-json-generate.mjs:19]: Yeah, we should probably try to do all in J's, so we can do this easily for other Debian bases distros.

File: lib/modules/versioning/ubuntu/distribution.ts (hunk 20 above)
Comment (Gabriel-Ladzaretti) [reply lib/modules/versioning/ubuntu/distribution.ts]: null/type-check related

File: tools/ubuntu-distro-json-generate.mjs (hunk 19 above)
Comment (Gabriel-Ladzaretti) [reply tools/ubuntu-distro-json-generate.mjs:19]: done both

File: data/debian-distro-info.json [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+  "1.2": "rex",
+  "1.3": "bo",
+  "2.0": "hamm",
+  "2.1": "slink",
+  "2.2": "potato",
+  "3.0": "woody",
+  "3.1": "sarge",
+  "4.0": "etch",
+  "5.0": "lenny",
+  "6.0": "squeeze"
```
Comment (rarkins) [inline data/debian-distro-info.json:18]: Was this file intended to be added to this PR?

Comment (JamieMagee) [reply data/debian-distro-info.json:18]: IMO it shouldn't be. I know I commented that we should cover Debian release names, but it should be a separate PR. There is some added complexity with Debian and it's naming of `stable`, `oldstable`, `oldoldstable`, etc.

Comment (Gabriel-Ladzaretti) [reply data/debian-distro-info.json]: removed the above

**Comment by Gabriel-Ladzaretti** [comment]
@viceice what is the right approach here,  using neither of these leads to null/type/eslint-check errors.
at least locally i guess 

> 🤔 [do we really need those fallbacks? Error should be caught on ci before publish](https://github.com/renovatebot/renovate/pull/14601/files/fecd0da745206e409012dceaf83273759a146177#r831544637)

and

>[ Why we need this? 🤔 Shouldn't be required.](https://github.com/renovatebot/renovate/pull/14601/files#r825645897)

**Review by rarkins — DISMISSED** [review]

**Review by viceice — APPROVED** [review]

**Review by viceice — CHANGES_REQUESTED** [review]
Can we add an entry to `package.json` scripts to call this?

File: tools/distro-json-generate.mjs [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+
+if (!shell.which(`sed`)) {
+  shell.echo('This script requires sed, exiting...');
+  shell.exit(2);
+}
+
+shell.echo(`OK`);
+
+const ubuntuDistroInfo = shell.exec(
+  `ubuntu-distro-info --all -f | sed -r 's/Ubuntu|"|LTS |Debian //g; s/([0-9]+.[0-9]+) /\\1 /; s/.*/\\L&/; s/( [a-z]*) [a-z]*/\\1/g; s/^[ \\t]*//'`,
```
[the remark this answers is not in the document]
Comment (Gabriel-Ladzaretti) [reply tools/distro-json-generate.mjs:19]: will be addressed in a future pr.

**Comment by Gabriel-Ladzaretti** [comment]
> Can we add an entry to `package.json` scripts to call this?

@viceice 
added but dropped the "generate:" prefix as its failing the ci builds

File: package.json [hunk 23]
```diff
@@ -21,6 +21,7 @@
     "eslint-fix": "eslint --fix . --report-unused-disable-directives",
     "generate": "run-s generate:*",
     "generate:imports": "node tools/generate-imports.mjs",
+    "ubuntu-distro-info": "node tools/distro-json-generate.mjs",
```
Comment (viceice) [inline package.json:24]:
```suggestion
    "update:ubuntu-distro-info": "node tools/distro-json-generate.mjs",
```

Let' suse `update:` prefix, please reorder alphanumeric.

Comment (Gabriel-Ladzaretti) [reply package.json]: done

**Review by viceice — APPROVED** [review]
